First proposal for a dplyr-database lesson, addressing issue 126

tomsing1 opened the pull request:
This is a first proposal with 07-dplyr-and-databases.Rmd, an alternative to / replacement of lesson `06-r-and-sql.Rmd` It addresses issue / contribution request #126 . I am happy to improve the proposal, if it is considered useful at all. Any feedback is very welcome.

### Some quick observations:

- I tried to stick to the existing lesson 06 as closely as possible, but using dplyr shifts the learning objectives from writing SQL to understanding the concept of lazy evaluation in R.
- The most difficult challenges of lesson 06 turned out to be very straightforward using dplyr's notation.
- There were some reference to a GUI (firefox plugin) to examine SQLite databases, which I believe are largely outdated by now. I have removed them, but am happy to add them back in if necessary. 
- The database creation section also leverages dplyr's `copy_to` command. Like the original version that directly used RSQLite, the example is not suitable for very large datasets, as it is funneled through R.
- Finally, in my opinion connecting to databases via dplyr feels like a much more natural next step 
given the previous lessons in this course. In this proposal, there is no need for learners to understand SQL, although an example of the translation that happens behind the scenes is provided.

fmichonneau commented:
Thanks for doing that @tomsing1! It looks great :smile: 

For this lesson, I would still like to show how to use the SQL syntax with dplyr to tie it back to the SQL lesson. It's useful to demonstrate that the dplyr syntax can directly be used on the database, but it seems also useful to show that you can send directly SQL queries to the database. So, it would be great if you could add a paragraph to show how to do that. It might also be interesting to contrast how much simpler it is to do the challenge using the pure dplyr syntax rather than mixing R and SQL to build the query.

We still use the firefox plugin to teach the SQL lesson, so it would be useful to keep references to it in the lesson.

Thanks again!

tomsing1 commented:
Thanks a lot for the feedback. I will implement the changes you suggested - great ideas!

fmichonneau commented:
I'm cleaning up issues, and there is one (#121) that indicates that we should include a link to download the SQL database from figshare.

Typically, this lesson is taught right after teaching SQL, so the learners have the file readily available (it's generated during the SQL lesson), but I can imagine that in some cases, the SQL lesson is skipped all together and the database isn't generated by the participants.

I'm going to add a note in your pull request, and it would be great if you could incorporate it. Thanks!

fmichonneau commented on 07-dplyr-and-databases.Rmd:
```diff
@@ -0,0 +1,340 @@
+already familiar with from previous lessons. Instead of reading the data from a 
+csv file, we will connect to an SQLite database with the same information
+(and more). The SQLite database is contained in a single file:
+`portal_mammals.sqlite` and we point R to it with dplyr's `src_sqlite()`
```
The SQLite database is contained in a single file `portal_mammals.sqlite` that that generated during the SQL lesson. If you don't have it, you can download it from Figshare using:

```{r download-database, eval=FALSE}
## download mammals database to data folder.
download.file("https://ndownloader.figshare.com/files/2292171", "data/portal_mammals.sqlite")
```
We can point R to this database with dplyr's `src_sqlite()` command.

tomsing1 commented:
Thanks a lot for your input and feedback @fmichonneau ! I have updated the lesson and incorporated your suggestions - except for referencing the GUI. I wasn't quite sure where that would fit in, especially as the lesson is pretty long already. 

In the current R-ecology curriculum, learners progress from using dplyr on local datasets to querying databases. So it seems they are already used to scripted analyses. Perhaps discussing the trade-offs of GUIs vs scripted queries could be done in the SQL lesson? Any feedback is very welcome!

fmichonneau commented:
That looks good! Thanks for doing this!
